examples/voting/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+    total_votes: VoteCount,
+}
+
+#[derive(Serial, Deserial, SchemaType)]
+struct VotingView {
+    description: Description,
+    tally:       Tally,
+    end_time:    Timestamp,
+}
+
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    description: Description,
+    ballots:     StateMap<AccountAddress, Vote, S>,
+    end_time:    Timestamp,
+}
+
+#[derive(Reject, Serial, PartialEq, Eq, Debug, SchemaType)]
+enum VotingError {
+    VotingFinished,
+    InvalidVoteIndex,
+    ParsingFailed,
+    ContractVoter,
+}
+
+impl From<ParseError> for VotingError {
+    fn from(_: ParseError) -> Self { VotingError::ParsingFailed }
+}
+
+type VotingResult<T> = Result<T, VotingError>;
+
+#[init(contract = "voting", parameter = "InitParameter")]
+fn init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+) -> InitResult<State<S>> {
+    let param: InitParameter = ctx.parameter_cursor().get()?;
+    Ok(State {
+        description: param.description,
+        ballots:     state_builder.new_map(),
+        end_time:    param.end_time,
+    })
+}
+
+#[receive(
+    contract = "voting",
+    name = "vote",
+    mutable,
+    parameter = "VoteParameter",
+    error = "VotingError"
+)]
+fn vote<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+) -> VotingResult<()> {
+    // Check that the election hasn't finished yet.
+    if ctx.metadata().slot_time() > host.state().end_time {
+        return Err(VotingError::VotingFinished);
+    }
+    // Ensure that the sender is an account.
+    let acc = match ctx.sender() {
+        Address::Account(acc) => acc,
+        Address::Contract(_) => return Err(VotingError::ContractVoter),
+    };
+    let param: VoteParameter = ctx.parameter_cursor().get()?;
+
+    let new_vote = param.vote;
+
+    // Check that vote is in range
+    ensure!(
+        (new_vote as usize) < host.state().description.options.len(),
+        VotingError::InvalidVoteIndex
+    );
+
+    // Insert or replace the vote for the account.
+    host.state_mut().ballots.entry(acc).and_modify(|vote| *vote = new_vote).or_insert(new_vote);
+    Ok(())
+}
+
+fn get_tally<S: HasStateApi>(
+    options: &[VotingOption],
+    ballots: &StateMap<AccountAddress, Vote, S>,
+) -> Tally {
+    let mut stats: BTreeMap<VotingOption, Vote> = BTreeMap::new();
+
+    for (_, ballot_index) in ballots.iter() {
+        let entry = &options[*ballot_index as usize];
+        stats.entry(entry.clone()).and_modify(|curr| *curr += 1).or_insert(1);
+    }
+    let total = stats.values().sum();
+
+    Tally {
+        result:      stats,
+        total_votes: total,
+    }
+}
+
+/// We assume that all ballots contain a valid voteoption index this should be
+/// checked by the vote function Assumption: Each account has at most one vote
+#[receive(contract = "voting", name = "getvotes", return_value = "VotingView")]
```

Comment:
Even if we accumulate the stats, this will not support a large number of VotingOptions.
Maybe a better design would be to have a "get top 10 query" and then also a getter for votes on a specific vote option.

Reply:
The votes are now accumulated in the `vote` function and saved in the `tally` state variable. There is a getter function `getNumberOfVotes` that retrieves the votes from the tally for a specific `vote option index`.

I added the following comment to the documentation because there is still a limitation in the size of the `Vec<VotingOption>`.

> //! `Vec<VotingOption>` (among other variables) is an input parameter to the
> //! `init` function as well as a return_value (among other variables) of the
> //! `view` function. Since there is a limit to the parameter/return_value size,
> //! the size of the `Vec<VotingOption>` is limited.

Reply:
I agree with this @DOBEN